src/server/twemcache-rs/src/protocol/data.rs

```diff
@@ -0,0 +1,672 @@
+// Copyright 2021 Twitter, Inc.
+// Licensed under the Apache License, Version 2.0
+// http://www.apache.org/licenses/LICENSE-2.0
+
+use crate::protocol::{CRLF, CRLF_LEN};
+
+use bytes::BytesMut;
+use segcache::Item;
+
+use core::convert::TryFrom;
+use std::borrow::Borrow;
+
+// TODO(bmartin):
+// - key length checking
+// - limit n_keys in get/gets
+// - limit total request length for writes (eg - segment size as an upper bound)
+//
+
+const NOREPLY: &[u8] = b"noreply";
+const NOREPLY_LEN: usize = NOREPLY.len();
+
+#[derive(PartialEq, Eq, Debug)]
+pub enum ParseError {
+    Incomplete,
+    Invalid,
+    UnknownCommand,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum MemcacheCommand {
+    Get,
+    Gets,
+    Set,
+    Add,
+    Replace,
+    Cas,
+    Delete,
+}
+
+impl TryFrom<&[u8]> for MemcacheCommand {
+    type Error = ParseError;
+
+    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+        let cmd = match value {
+            b"get" => MemcacheCommand::Get,
+            b"gets" => MemcacheCommand::Gets,
+            b"set" => MemcacheCommand::Set,
+            b"add" => MemcacheCommand::Add,
+            b"replace" => MemcacheCommand::Replace,
+            b"cas" => MemcacheCommand::Cas,
+            b"delete" => MemcacheCommand::Delete,
+            _ => {
+                return Err(ParseError::UnknownCommand);
+            }
+        };
+        Ok(cmd)
+    }
+}
+
+pub trait Request {
```

Comment:
Add a TODO indicating that this will eventually be lifted out and shared with other protocols